discovery/discovery_test.go

```diff
@@ -96,18 +98,59 @@ func TestNewDBOrchestratorPoolCache_GivenListOfOrchs_CreatesPoolCacheCorrectly(t
 	require.NotNil(dbOrch)
 }
 
+type stubRpc struct {
+}
+
+func (s *stubRpc) getOrchestrator(ctx context.Context, orchestratorServer *url.URL) (*net.OrchestratorInfo, error) {
+	return &net.OrchestratorInfo{Transcoder: orchestratorServer.String()}, nil
+}
+
+func TestTempGetOrchestrators_GivenOneOrchestratorRequested_ReturnsRandomOrch(t *testing.T) {
+	dbh, dbraw, err := common.TempDB(t)
+	defer dbh.Close()
+	defer dbraw.Close()
+	require := require.New(t)
+	assert := assert.New(t)
+	require.Nil(err)
+
+	addresses := []string{"https://127.0.0.1:8936", "https://127.0.0.1:8937", "https://127.0.0.1:8938"}
+
+	node, _ := core.NewLivepeerNode(nil, "", nil)
+	orchPool := NewOrchestratorPool(node, addresses)
+	orchPool.rpc = &stubRpc{}
+
+	orchInfo, err := orchPool.GetOrchestrators(1)
+	require.Nil(err)
+	assert.Len(orchInfo, 1)
+
+	secondOrchPool := NewOrchestratorPool(node, addresses)
+	secondOrchPool.rpc = &stubRpc{}
+
+	secondOrchInfo, err := secondOrchPool.GetOrchestrators(1)
+	require.Nil(err)
+	assert.Len(secondOrchInfo, 1)
+
+	thirdOrchPool := NewOrchestratorPool(node, addresses)
+	thirdOrchPool.rpc = &stubRpc{}
+
+	thirdOrchInfo, err := thirdOrchPool.GetOrchestrators(1)
+	require.Nil(err)
+	assert.Len(thirdOrchInfo, 1)
+
+	if orchInfo[0].Transcoder == secondOrchInfo[0].Transcoder && secondOrchInfo[0].Transcoder == thirdOrchInfo[0].Transcoder && orchInfo[0].Transcoder != thirdOrchInfo[0].Transcoder {
+		t.Error("Expected list of orchAddresses returned to be randomized")
+	}
+}
+
 func TestNewOrchestratorPoolCache_GivenListOfOrchs_CreatesPoolCacheCorrectly(t *testing.T) {
 	node, _ := core.NewLivepeerNode(nil, "", nil)
 	addresses := []string{"https://127.0.0.1:8936", "https://127.0.0.1:8937", "https://127.0.0.1:8938"}
 	expectedOffchainOrch := StubOrchestratorPool(addresses)
 	assert := assert.New(t)
-	require := require.New(t)
 
 	// creating NewOrchestratorPool with orch addresses
 	offchainOrch := NewOrchestratorPool(node, addresses)
-	for i, uri := range offchainOrch.uris {
-		assert.Equal(uri.String(), expectedOffchainOrch.uris[i].String())
-	}
+	assert.Len(offchainOrch.uris, 3)
```

Comment:
Seed the default RNG with a constant value before this point (but after LivepeerNode is initialized). Then see what order the O list is initialized. Update the expected offchain orch list to match. That preserves the character of this test -- just checking for lengths doesn't offer as strong of an assurance.

```diff
diff --git a/discovery/discovery_test.go b/discovery/discovery_test.go
index c5c5d979..78f4aebf 100644
--- a/discovery/discovery_test.go
+++ b/discovery/discovery_test.go
@@ -2,6 +2,7 @@ package discovery
 
 import (
 	"context"
+	"math/rand"
 	"net/url"
 	"testing"
 
@@ -145,29 +146,25 @@ func TestTempGetOrchestrators_GivenOneOrchestratorRequested_ReturnsRandomOrch(t
 func TestNewOrchestratorPoolCache_GivenListOfOrchs_CreatesPoolCacheCorrectly(t *testing.T) {
 	node, _ := core.NewLivepeerNode(nil, "", nil)
 	addresses := []string{"https://127.0.0.1:8936", "https://127.0.0.1:8937", "https://127.0.0.1:8938"}
-	expectedOffchainOrch := StubOrchestratorPool(addresses)
+	expected := []string{"https://127.0.0.1:8938", "https://127.0.0.1:8937", "https://127.0.0.1:8936"}
 	assert := assert.New(t)
 
 	// creating NewOrchestratorPool with orch addresses
+	rand.Seed(321) // reset rng
 	offchainOrch := NewOrchestratorPool(node, addresses)
-	assert.Len(offchainOrch.uris, 3)
 
 	// creating new OrchestratorPool with different first value
-	addresses[0] = "https://127.0.0.1:89"
-	expectedOffchainOrch = StubOrchestratorPool(addresses)
-	assert.NotEqual(offchainOrch.uris[0].String(), expectedOffchainOrch.uris[0].String())
+	for i, uri := range offchainOrch.uris {
+		assert.Equal(uri.String(), expected[i])
+	}
 
 	orchestrators := StubOrchestrators(addresses)
 	node.Eth = &eth.StubClient{Orchestrators: orchestrators}
 
 	// testing NewOnchainOrchestratorPool
+	rand.Seed(321) // reset rng
 	offchainOrchFromOnchainList := NewOnchainOrchestratorPool(node)
-	newAddressExists := false
-	for _, uri := range offchainOrchFromOnchainList.uris {
-		if uri.String() == addresses[0] {
-			newAddressExists = true
-		}
+	for i, uri := range offchainOrchFromOnchainList.uris {
+		assert.Equal(uri.String(), expected[i])
 	}
-
-	assert.True(newAddressExists)
 }
``` 

With this, I don't believe `TestTempGetOrchestrators_GivenOneOrchestratorRequested_ReturnsRandomOrch` is needed anymore since we're also testing a randomized list as part this function.

Perhaps a slightly saner method would be to make the permutation monkey-patchable. Have tests use a monkey-patched function returning a fixed sequence (say the original input without any shuffling) and the actual RNG can be tested in isolation, in a manner similar to what was outlined above.

Reply:
Interesting! Thank you for the suggestion. Setting RNG with a constant value makes the test truly deterministic, and allows for checking of randomness simultaneously. No need to monkey-patch for the other test.